### ISSUE TITLE
[$1000] If user was removed from workspace while he is offline he still has access to #announce

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#21519]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
Use 2 accounts (UserA and UserB) on different browsers same time
1. as UserA, create workspace
2. as UserA, add userB to workspace
3. as UserB, open #announce Report
4. as UserA, remove UserB from workspace
5. as UserB, you will see screen that you don't have access to this chat (it's expected behaviour).
6. as UserA, add UserB again to workspace
7. as UserB, close browser
8. as UserA, remove UserB from workspace
9. as UserB, open website again and you will see #announce chat with badge 'Unavailable workspace'.
10. as UserB, post message to this #announce chatroom and you will see issue "Unexpected error while posting the comment, please try again later
Auth CreateReportAction returned an error"
## Expected Result:
Removed user should not see #announce chat of unavailable workspace. And if he will go by direct link he should get same see screen that he doesn't have access to this chat
## Actual Result:
If user was removed from workspace while he is offline he will keep access to #announce chatroom
## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android / native
- [x] Android / Chrome
- [x] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** 1.3.32-5
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/ad8cd8cc-3e9b-453b-baf5-5f32d95a190c

https://github.com/Expensify/App/assets/43996225/20a5be04-c66d-4037-b314-ad1ce691101e

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @mkarkachov
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1686850018180709

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01208945f2956c780b</li>
        <li>Upwork Job ID: 1673798077535150080</li>
        <li>Last Price Increase: 2023-07-04</li>
    </ul>
</details>

**Comment by mkarkachov** [comment]
video from original [bug report ](https://expensify.slack.com/archives/C049HHMV9SM/p1686850018180709)

https://github.com/Expensify/App/assets/1360242/e646f61c-894e-4206-a5d0-09696b3551ba

**Comment by alexpensify** [comment]
Updating to be `External`. I confirmed that this issue happens and applies when the admin deletes the workspace too! 

![image](https://github.com/Expensify/App/assets/8107070/046d5013-e5af-4d5d-8119-b60cd305b2f1)

**Comment by rain2o** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

If a user is removed from a workspace, or a workspace is deleted, while that user is offline, they are still able to access some rooms from that workspace.

### What is the root cause of that problem?

**TLDR:** Reports are removed when this action happens while subscribed to Pusher updates, but not when loading updates on a new app load.

**Detailed explanation**
While the browser is open we subscribe to Pusher updates. When a user is removed from a workspace their browser receives pusher updates to delete that policy and its related reports. Below is an example of such an update:
```json
[{"eventType":"onyxApiUpdate","data":[{"onyxMethod":"set","key":"report_827406856126170","value":null}]}]
```
When this happens we execute each of those onyx methods, which deletes the reports from storage.

If a user is removed from a workspace while they do not have Expensify active, these pusher updates are not received. The next time they open the app and the policies are updated, we are only updating the reports which belong to that policy to no longer have a draft, rather than removing them like we do in the above scenario.
```javascript
_.each(policyReports, ({reportID}) => {
    cleanUpMergeQueries[`${ONYXKEYS.COLLECTION.REPORT}${reportID}`] = {hasDraft: false};
    cleanUpSetQueries[`${ONYXKEYS.COLLECTION.REPORT_DRAFT_COMMENT}${reportID}`] = null;
});
Onyx.mergeCollection(ONYXKEYS.COLLECTION.REPORT, cleanUpMergeQueries);
Onyx.multiSet(cleanUpSetQueries);
```

### What changes do you think we should make in order to solve the problem?


We should change the [policy subscription callback](https://github.com/Expensify/App/blob/main/src/libs/actions/Policy.js#L21) to delete any reports belonging to a policy when its value is null, to be consistent with the Pusher updates received.
```javascript
const cleanUpSetQueries = {};
_.each(policyReports, ({reportID}) => {
    cleanUpSetQueries[`${ONYXKEYS.COLLECTION.REPORT}${reportID}`] = null;
});
Onyx.multiSet(cleanUpSetQueries);
```

### What alternative solutions did you explore? (Optional)

I noticed this comment in the above mentioned callback:
> // If we are deleting a policy, we have to check every report linked to that policy
> // and unset the draft indicator (pencil icon) alongside removing any draft comments. Clearing these values will keep the newly archived chats from being displayed in the LHN.
> // More info: https://github.com/Expensify/App/issues/14260

If this is expected behavior, to keep the reports in the LHN and simply change it to say `Unavailable workspace`, then we will need to update the logic in (ReportScreen.js)[https://github.com/Expensify/App/blob/main/src/pages/home/ReportScreen.js#L269] to show the `FullPageNotFoundView` in this scenario. Something along these lines:

```javascript
// src/libs/ReportUtils.js
function canAccessReport(report) {
    return lodashGet(allPolicies, `${ONYXKEYS.COLLECTION.POLICY}${report.policyID}`) !== undefined;
}
```

```javascript
// src/pages/home/ReportScreen.js
<FullPageNotFoundView
    shouldShow={(!this.props.report.reportID && !this.props.report.isLoadingReportActions && !isLoading)
        || shouldHideReport
        || !ReportUtils.canAccessReport(this.props.report)
    }
...
```

**Comment by alexpensify** [comment]
@mananjadhav - when you get a chance, can you please review this proposal? Thank you!

**Comment by alexpensify** [comment]
Reassigning another `bug` team member, I'm going OOO until Monday, July 10, and will take it back then if it's not closed by then. 

Required action from the 🐛 team: Make sure the proposal process is moving forward, so we can get this one assigned out to fix.

**Comment by mananjadhav** [comment]
I'll be looking at these today.

**Comment by johncschuster** [comment]
@mananjadhav sorry for the additional bump. Do you feel we have what we need? If not, let's take this to #expensify-open-source.

**Comment by rain2o** [comment]
> @mananjadhav sorry for the additional bump. Do you feel we have what we need? If not, let's take this to #expensify-open-source.

If my proposal isn't acceptable I'm happy to dig into it more and find a better solution. Just need to know if that's the case.

**Comment by mananjadhav** [comment]
Apologies I couldn't check this as I was OOO. I tried looking at the logic and I also saw the comments. But I think it's best if @s77rt / @rushatgabhane picks this, considering my limited bandwidth and they've worked on this earlier.

**Comment by s77rt** [comment]
I'd like to take this as C+

**Comment by alexpensify** [comment]
Alright, removing the other 🐛 team member. I'm back online and see that this is our current state: 

@s77rt - thanks! I've assigned you as the C+ here. Could you please review the proposal and share any remaining feedback. Thank you!

**Comment by s77rt** [comment]
@rain2o Thanks for the proposal. I don't think your RCA is complete/accurate. We are indeed missing onyx updates here but the code you linked is unrelated, the draft and the draft status are cleared that way because those two props are client-side only while in this issue we are dealing with server driven data. I noticed that calling `OpenReport` will return a `jsonCode=403` with no onyx data, the only onyx data being applied here is the one we set as `failureData` on `openReport` function.

**Comment by s77rt** [comment]
I asked on Slack https://expensify.slack.com/archives/C01GTK53T8Q/p1689031771678159 for some guidance on the expected response of `OpenReport` call in such cases.

**Comment by mkarkachov** [comment]
> 📣 @mkarkachov We're missing your Upwork ID to automatically send you an offer for the Reporter role. Once you apply to the [Upwork job](https://www.upwork.com/jobs/~01208945f2956c780b), your Upwork ID will be stored and you will be automatically hired for future jobs!

Have just applied.

**Comment by rain2o** [comment]
> @rain2o Thanks for the proposal. I don't think your RCA is complete/accurate. We are indeed missing onyx updates here but the code you linked is unrelated, the draft and the draft status are cleared that way because those two props are client-side only while in this issue we are dealing with server driven data. I noticed that calling `OpenReport` will return a `jsonCode=403` with no onyx data, the only onyx data being applied here is the one we set as `failureData` on `openReport` function.

Hmm, interesting. If we depend on the response of `OpenReport`, then each and every room from which the user has been removed will remain in the LHN until they open each one I believe (unless OpenReport is called in the background on all of them). That seems like a frustrating UX, especially if the user has been removed from a workspace with many rooms.

I proposed those changes because while testing I found upon performing step 9, the policy collection was updated, and the value for the policy ID to which those reports belonged was now set as `null`. As a result, I thought it would make sense to set any reports to `null` which belong to those nulled Policies. Perhaps I'm wrong in that assumption.

I will continue to investigate.

**Comment by rain2o** [comment]
Looking into the behavior a bit more, it looks like the reason the Policies collection is updated is because of the response we receive from the `ReconnectApp` request when opening the app again. The request sent the list of policyIDs including the policy from which the user has been removed (because the client doesn't know this yet). The response returned
```
{
    "onyxMethod": "set",
    "key": "policy_103AE0F538586863",
    "value": null
}
```

Which triggers the behavior in my proposal.

Alternatively, we could request that the response from `ReconnectApp` includes onyx methods to set all Reports as `null` to which the user no longer has access. In that case this issue would need to become internal for backend changes.


You said that my changes were unrelated. Could you help me understand how/why? To me, it seems like a direct effect of the Policy being removed. Perhaps I'm misunderstanding policies, but my understanding is that the policy is what determines what the user has access to, in this case these workspace reports. I don't really understand the behavior described in the comment I linked to.

>If we are deleting a policy, we have to check every report linked to that policy and unset the draft indicator (pencil icon) alongside removing any draft comments. Clearing these values will keep the newly archived chats from being displayed in the LHN.

If we're deleting a policy, why would we keep the reports which belong to that policy? What am I missing?

**Comment by s77rt** [comment]
@rain2o Thanks for the update.

> Alternatively, we could request that the response from ReconnectApp includes onyx methods to set all Reports as null to which the user no longer has access. In that case this issue would need to become internal for backend changes.

That actually seems to be a better idea :+1:.

> You said that my changes were unrelated. Could you help me understand how/why?

What I meant is that we should not couple such data this way. That code is only used for draft and draft status because those data are not sent by the server. That code is more like a cleanup code for client-only data.

> If we're deleting a policy, why would we keep the reports which belong to that policy? 

This is used for archived reports. If you delete your own workspace, the reports will be archived instead of removed and you'd still have access to.

**Comment by s77rt** [comment]
@alexpensify Can we get an engineer opinion here? As @rain2o stated the bug seems internal where the `ReconnectApp` API response deletes old policies that you had access to, but does not delete their linked reports.

**Comment by rain2o** [comment]
Thanks for the explanations @s77rt , that helps. I didn't realize we archive reports in that way.

**Comment by alexpensify** [comment]
Will do, I'll start a discussion tomorrow.

**Comment by alexpensify** [comment]
I've run out of time and will start a discussion soon.

**Comment by alexpensify** [comment]
This one is on my radar this week.

**Comment by alexpensify** [comment]
I've started the discussion here: https://expensify.slack.com/archives/C01GTK53T8Q/p1689625060516219

**Comment by alexpensify** [comment]
I bumped the conversation again

**Comment by alexpensify** [comment]
@luacmartins - I'm reassigning the `engineering` label because my previous bumps didn't get a response yet. We need help to review @s77rt from here: 

https://github.com/Expensify/App/issues/21519#issuecomment-1631553935

Thank you! Your response will help us plan for the next steps here.

**Comment by luacmartins** [comment]
I don't think we should delete those reports. We instead archive them for removed users so that users still have access to the history of the conversation, but can no longer post comments to it.

**Comment by s77rt** [comment]
@luacmartins The reports become archived only for the admins but for the rest they lose complete access. It looks the current behaviour is not the intended one? Either way this seems internal, right?

**Comment by alexpensify** [comment]
@luacmartins - when you get a chance, can you review @s77rt most recent question? Thanks!

**Comment by alexpensify** [comment]
@luacmartins - I believe you are back online tomorrow, please review @s77rt question here: 

https://github.com/Expensify/App/issues/21519#issuecomment-1644783195

Thank you!

**Comment by luacmartins** [comment]
@s77rt I think the chat is archived for both. See image below, admin on left and removed employee on right:

![Screenshot 2023-07-25 at 10 29 20 AM](https://github.com/Expensify/App/assets/22219519/ba217a10-3bc5-4e25-a1a9-65989b13e17a)

So I don't think we should remove the report for the user, but we need to restrict posting new messages.

**Comment by luacmartins** [comment]
As if this is internal or not, do we know the root cause yet?

**Comment by s77rt** [comment]
@luacmartins

- Online: After you get removed from a WS, the report is completely removed <-- This is coming from the pusher event that set the report to null. 
- Offline: In case you missed the pusher events, the next time you call `OpenReport` it will return onyx update `report_id: null` and clears the report (this response is only returned when you don't have access to a report)



https://github.com/Expensify/App/assets/16493223/320ee7d0-73b1-4a05-81b6-935724c5d873

**Comment by luacmartins** [comment]
Got it! Thanks! In that case this is internal. I'll take a look at it some time this week.

**Comment by alexpensify** [comment]
@luacmartins - any update here? Thanks!

**Comment by luacmartins** [comment]
Not yet! I've been focused on other wave3 and wave4 issues this week. I'll work on this next week.

**Comment by luacmartins** [comment]
@flodnv I see that you added [this code](https://github.com/Expensify/Web-Expensify/blob/f683cb2d1e115ef7b3b8ccde9b013166c7260df7/lib/PolicyAPI.php#L4661-L4671) to set report and reportActions = null if the user is removed from a Workspace, but that behavior is introducing this bug. Did we change the behavior to no longer archive chats and just remove them completely?

**Comment by flodnv** [comment]
AFAIK we do not archive #announce 😕 When you get removed from the policy you should get removed from #announce, no?

**Comment by flodnv** [comment]
>If user was removed from workspace while he is offline he still has access to #announce

I could not reproduce this bug when I set out to fix it 😢 https://expensify.slack.com/archives/C01GTK53T8Q/p1686076334288629?thread_ts=1684863673.526879&cid=C01GTK53T8Q

**Comment by luacmartins** [comment]
Oh I think you're right and we just archive user created policy rooms, not default rooms. Thanks for the link to that thread. It seems like we had aligned on option #2, which is to keep the chat in the LHN and remove it when opening the report. That seems to be the current behavior, since whenever I navigated to the chat, it was removed and I never managed to send a message.

TBH I think what we have here might be the expected behavior. Would you agree @flodnv?

**Comment by flodnv** [comment]
Based on what you said and what I can see in the video in the OP, I agree with you.

**Comment by luacmartins** [comment]
Sounds good to me. I think we can :donothing: and close this issue then, since it's expected.

**Comment by alexpensify** [comment]
Thank you for the update! I've closed the job in Upwork too!

**Comment by mkarkachov** [comment]
Hi guys, it's not clear to me actually. Will I get bug bounty for reporting this issue? @alexpensify @luacmartins

**Comment by alexpensify** [comment]
@mkarkachov - Thanks for reaching out. In this case, there was no action to fix the issue and it was decided that it was expected behavior. With that said, there is no reporting bounty because there was no PR applied to this one.
